afpacket/afpacket.go

```diff
@@ -212,18 +214,29 @@ func (h *TPacket) setUpRing() (err error) {
 	return nil
 }
 
-// Close cleans up the TPacket.  It should not be used after the Close call.
-func (h *TPacket) Close() {
-	if h.fd == -1 {
-		return // already closed.
-	}
+// Finalize closes the socket (if needed) and also unmaps the ring buffer.
```

Comment:
keep this unexported and just export Close?

Reply:
I put this through extensive testing and found that the background GC might not reclaim fast enough if I open/read/close in a tight loop, and eventually the OS would kill the process (probably OOM). Making an explicit call to `Finalize()` fixes it.

I'm also thinking about what happens if calling code keeps a reference to previously read data even after the `TPacket` goes out of scope.

Both of these concerns seem unlikely, but I felt that exporting the method (and its godoc) made it clearer that these problems might exist, and gave the implementer a better understanding and control over the validity and life cycle of the data buffer.

Of course, it's still a strange, hacky thing to expose. So if it's better to hide the method, I'll see if we can have the read path at least attempt to handle the cleanup itself. I didn't investigate, but it might not be that difficult...

Reply:
Actually, it turned out pretty easy to just call `finalize()` if the read fails and the FD is closed. That _mostly_ works, so I'll push a commit when my tests finish.

However, I'll note that my test harness is around production code that uses a `context.Context` to trigger both the `Close()` and also to exit the read loop. (This seems pretty representative of a real use case.) And sometimes, the read loop exits early because it sees the `context` is done, and not because of an actual read failure. In this case, we'd need to make _one final_ call to the read method to make sure the buffer gets finalized. If I don't do that, my stress test loop will still eventually get killed because the GC isn't finalizing and un-mapping buffers fast enough.